### ISSUE TITLE
Blowing up a huge amount of TNT causes performance fail

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest_game#3010]
If you blow up a HUGE amount of TNT (like, a 100×100×100 cube), the game quickly slows down to a crawl after the first explosion.

**Comment by appgurueu** [comment]
I'd say this is roughly to be expected. TNT explosions simply are expensive. Even if you got them down to 1ms per explosion, blowing up a million (100³) TNT nodes at roughly the same time is expected to take a thousand seconds.

That said, there very likely is potential for optimization (even though such "HUGE" amounts will always slow the game down to a crawl). Might take a look using the LuaJIT profiler eventually.

Some options I see:

* Optimize TNT. This won't suffice to fix this issue.
* Add TNT explosion scheduling, delay explosions if a budget is exceeded. Will keep the game playable at the cost of the explosions being, well, delayed, which is inaccurate. Your million TNT nodes would be exploding for *a while*.
* Do some batching of nearby TNT explosions (likely nontrivial). *Might* suffice to fix this issue without a "compromise".

What would you suggest?

**Comment by Wuzzy2** [comment]
I have no suggestions yet, but there is one other thing to look at:

* If there is a high number of particles, client comes down to a crawl (this is a separate issue, I think)

But yeah I believe this is hard to do right, especially due to possible exponential growth. :-( I also agree with low priority; I don't really care that much about the bug but I posted it anyway for documentation purposes.

**Comment by appgurueu** [comment]
> If there is a high number of particles, client comes down to a crawl (this is a separate issue, I think)

Yes, mostly an engine issue (although again "batching" as described by Desour could be seen as mod responsibility)

**Comment by tenplus1** [comment]
Agree'd, massive tnt explosions can bring any system to a crawl which is why I limited particles and disabled some non-essential functions for it to work a little more smoothly for my own fork: https://notabug.org/tenplus1/tnt

**Comment by Destroyinator69420** [comment]
I could see a flag in the game that when a large enough amount of TNT is detonated, then the game will skip the visuals of the detonation and skip straight to the result. Mojang did this in minecraft pi edition as well as early versions of minecraft pocket edition too.

**Comment by SmallJoker** [comment]
The TNT mod could make better use of ...

1. VManip: by recursively collecting all TNT in range to then get one VManip object of the entire affected volume to then process each TNT position.
2. Protection checks by using `minetest.is_area_protected` within `radius` (i.e. cube that encloses the sphere) to reduce the `minetest.is_protected` calls roughly by the factor 2³ = 8.

As soon the affected volume size is known, such performance optimizations could be made at ease.
